### PR TITLE
Allow staff to use ‘real’ names as well

### PR DESCRIPTION
**Pull request by AaronVanGeffen**
Screenshot 2024-09-09 at 09 28 29

### CODE REVIEW

File: src/openrct2/entity/Peep.cpp [hunk 1]
```diff
@@ -1611,7 +1612,7 @@ void Peep::FormatNameTo(Formatter& ft) const
             ft.Add<StringId>(_staffNames[staffNameIndex]);
             ft.Add<uint32_t>(PeepId);
         }
-        else if (GetGameState().Park.Flags & PARK_FLAGS_SHOW_REAL_GUEST_NAMES)
+        if (gameState.Park.Flags & PARK_FLAGS_SHOW_REAL_GUEST_NAMES)
```
Comment (AaronVanGeffen) [inline src/openrct2/entity/Peep.cpp:1615]: Probably should keep this as an `else if`.

File: src/openrct2/config/Config.cpp [hunk 2]
```diff
@@ -237,6 +237,7 @@ namespace OpenRCT2::Config
             model->RenderWeatherGloom = reader->GetBoolean("render_weather_gloom", true);
             model->ShowGuestPurchases = reader->GetBoolean("show_guest_purchases", false);
             model->ShowRealNamesOfGuests = reader->GetBoolean("show_real_names_of_guests", true);
+            model->ShowRealNamesOfStaff = reader->GetBoolean("show_real_names_of_staff", true);
```
Comment (AaronVanGeffen) [inline src/openrct2/config/Config.cpp:240]: Not sure if this should be `true` by default. Personally, I prefer to have it off, I think.

File: src/openrct2-ui/windows/Options.cpp [hunk 3]
```diff
@@ -377,13 +378,14 @@ namespace OpenRCT2::Ui::Windows
...
-        MakeWidget({175, kTweaksStart + 61}, {125, 12}, WindowWidgetType::DropdownMenu, WindowColour::Secondary                                                                      ), // Default inspection time dropdown
-        MakeWidget({288, kTweaksStart + 62}, { 11, 10}, WindowWidgetType::Button,       WindowColour::Secondary, STR_DROPDOWN_GLYPH,       STR_DEFAULT_INSPECTION_INTERVAL_TIP       ), // Default inspection time dropdown button
+        MakeWidget({  5,  kTweaksStart + 0}, {300, 96}, WindowWidgetType::Groupbox,     WindowColour::Secondary, STR_OPTIONS_TWEAKS                                                  ),
+        MakeWidget({ 10, kTweaksStart + 15}, {290, 15}, WindowWidgetType::Checkbox,     WindowColour::Tertiary , STR_REAL_NAME_GUESTS,     STR_REAL_NAME_GUESTS_TIP                  ), // Show 'real' names of guests
+        MakeWidget({ 10, kTweaksStart + 30}, {290, 15}, WindowWidgetType::Checkbox,     WindowColour::Tertiary , STR_REAL_NAME_STAFF,      STR_REAL_NAME_STAFF_TIP                   ), // Show 'real' names of guests
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/Options.cpp:383]:
```suggestion
        MakeWidget({ 10, kTweaksStart + 30}, {290, 15}, WindowWidgetType::Checkbox,     WindowColour::Tertiary , STR_REAL_NAME_STAFF,      STR_REAL_NAME_STAFF_TIP                   ), // Show 'real' names of staff
```

File: src/openrct2/entity/Peep.cpp [hunk 4]
```diff
@@ -2654,19 +2655,20 @@ int32_t PeepCompare(const EntityId sprite_index_a, const EntityId sprite_index_b
...
         gameState.Park.Flags &= ~PARK_FLAGS_SHOW_REAL_GUEST_NAMES;
-        // Peep names are now dynamic
-    }
+
+    if (config.ShowRealNamesOfStaff)
+        gameState.Park.Flags |= PARK_FLAGS_SHOW_REAL_STAFF_NAMES;
+    else
+        gameState.Park.Flags &= ~PARK_FLAGS_SHOW_REAL_STAFF_NAMES;
 
     auto intent = Intent(INTENT_ACTION_REFRESH_GUEST_LIST);
```
Comment (tupaschoal) [inline src/openrct2/entity/Peep.cpp:2673]: Does the intent refresh everything or just the guest as the name implies? Because if it's everything, we can follow-up on  another PR to break down and send smaller intents, but if it's just the guest list, then this is not refreshing what it should, right?

File: src/openrct2/entity/Peep.cpp (hunk 1 above)
Comment (tupaschoal) [reply src/openrct2/entity/Peep.cpp:1615]: Yeah, I think you should start with the real guest name here, maybe:
```c++
if (gameState.Park.Flags & (PARK_FLAGS_SHOW_REAL_GUEST_NAMES || PARK_FLAGS_SHOW_REAL_STAFF_NAMES))
{
    auto realNameStringId = GetRealNameStringIDFromPeepID(PeepId);
    ft.Add<StringId>(realNameStringId);
}
else if (auto* staff = As<Staff>(); staff != nullptr)
{
    // ...
}
else
{
    ft.Add<StringId>(STR_GUEST_X).Add<uint32_t>(PeepId);
}
```

File: src/openrct2/entity/Peep.cpp (hunk 4 above)
Comment (AaronVanGeffen) [reply src/openrct2/entity/Peep.cpp:2673]:
```cpp
            case INTENT_ACTION_REFRESH_GUEST_LIST:
                WindowGuestListRefreshList();
                break;
```

It's only forcing the guest list to refresh. I'm not seeing any issues with the staff list, though. Perhaps it isn't cached like the guest list? I've not looked into it much.

Comment (tupaschoal) [reply src/openrct2/entity/Peep.cpp:2673]: That seems correct, there is also this comment on StaffList:
https://github.com/OpenRCT2/OpenRCT2/blob/e6e3f5db7696697337b43235d08633beeffe444c/src/openrct2-ui/windows/StaffList.cpp#L214-L216

Maybe we should split the functions just to avoid refreshing the guest list unnecessarily, then?

Comment (AaronVanGeffen) [reply src/openrct2/entity/Peep.cpp:2673]: I don't think it's worth it, tbh. This is only called once, when the option is toggled, after all.

Comment (tupaschoal) [reply src/openrct2/entity/Peep.cpp:2673]: Your call

File: src/openrct2/entity/Peep.cpp (hunk 1 above)
Comment (AaronVanGeffen) [reply src/openrct2/entity/Peep.cpp:1615]: While I like the looks of that, it doesn't quite work, unfortunately. You need an explicit check for staff type, otherwise you'd get a real name for staff with just `PARK_FLAGS_SHOW_REAL_GUEST_NAMES` set.

File: src/openrct2/entity/Peep.cpp [hunk 5]
```diff
@@ -1611,7 +1612,7 @@ void Peep::FormatNameTo(Formatter& ft) const
             ft.Add<StringId>(_staffNames[staffNameIndex]);
             ft.Add<uint32_t>(PeepId);
         }
-        else if (GetGameState().Park.Flags & PARK_FLAGS_SHOW_REAL_GUEST_NAMES)
+        else if (gameState.Park.Flags & PARK_FLAGS_SHOW_REAL_GUEST_NAMES)
```
Comment (AaronVanGeffen) [inline src/openrct2/entity/Peep.cpp:1615]: This won't work if staff have real names, but guests don't.

**Comment by GingerAdonis** [comment]
Looks like the whole family works in your park. 🥲

**Comment by AaronVanGeffen** [comment]
> Looks like the whole family works in your park. 🥲

C. is for Cleaner? 🥲

File: src/openrct2/config/Config.cpp (hunk 2 above)
Comment (tupaschoal) [reply src/openrct2/config/Config.cpp:240]: I prefer to have it off too, while peeps names are just often random information, one might be using the staff numbers to memorize where they're assigned to or whatever.

It also should be `false` given that the changelog mentions it is optional

**Review by tupaschoal — APPROVED** [review]
I haven't tested, but otherwise LGTM, save for a minor detail on the config default.

File: src/openrct2/Context.cpp [hunk 6]
```diff
@@ -848,9 +849,7 @@ namespace OpenRCT2
                     }
 #endif
                 }
-                // This ensures that the newly loaded save reflects the user's
```
Comment (Gymnasiast) [inline src/openrct2/Context.cpp:851]: Why did you move this?

Comment (AaronVanGeffen) [reply src/openrct2/Context.cpp:851]: It wasn't necessary, but I thought it cleaner to have the logic all _before_ the network logic. Ideally, the network logic will be split off to its own function. As I was touching the line anyway, I thought this was as good a time as any.

Comment (Gymnasiast) [reply src/openrct2/Context.cpp:851]: I’m just afraid it will cause desyncs, so I’d rather it gets moved down again. Apart from this, the PR looks good to me.

Comment (AaronVanGeffen) [reply src/openrct2/Context.cpp:851]: I'm reasonably confident it won't cause a desync, but I don't feel strongly about the move. I'll move it back down.

**Review by Gymnasiast — APPROVED** [review]

**Comment by sta5is** [comment]
Great addition! I notice two issues with reused names:

1) The same names are reused for different types of staff. For instance, the first handyman will be named Bernard C. and the first mechanic is also named the same. This also applies to the security guard and entertainer staff. Subsequently, the next name, Lindsey C. is also used for all the second staff on the list.

2) The same names are always used. First staff is always named Bernard, second is Lindsey, third is Cynthia, and etc. I tested with different scenarios and the names used are in that same ordered. Would it be possible to assign their names more randomly?

**Comment by AaronVanGeffen** [comment]
@sta5is These are good suggestions. I don't have time to pick this up now, but I'd like to take care of them. Could you file an issue for this, so we don't lose track of it? Thanks.

**Comment by MASHzj** [comment]
> Great addition! I notice two issues with reused names:
> 
> 1. The same names are reused for different types of staff. For instance, the first handyman will be named Bernard C. and the first mechanic is also named the same. This also applies to the security guard and entertainer staff. Subsequently, the next name, Lindsey C. is used for all the second staff of each type.
> 2. The same names are always used. First staff is always named Bernard, second is Lindsey, third is Cynthia, and etc. I tested with different scenarios and the names used are in that same ordered. Would it be possible to assign their names more randomly?

I know from a video of Marcel Vos that the names for guests are always in the same order as well.

**Comment by sta5is** [comment]
@AaronVanGeffen Sorry for the delay. Opened #22809 for issue tracking.
